theano/gpuarray/sort.py

```diff
@@ -0,0 +1,314 @@
+                dsti_strides=dsti_strides_code if self.return_indices else '',
+                src_strides=src_strides_code,
+                set_slice=set_slice_code,
+                write_value=int(self.return_values),
+                write_index=int(self.return_indices),
+                ndim=str(ndim),
+                use_half=int(node.inputs[0].dtype == 'float16')
+                )
+        elif device_type == b'opencl':
+            raise NotImplementedError()
+
+        # setup parameters
+        param_types = [ga.SIZE] * (ndim - 1)  # dims
+        for _ in range(self.return_values + self.return_indices):
+            param_types.append(ga.GpuArray)  # dst*
+            param_types.extend([ga.SSIZE] * ndim)  # dst*_strides
+        param_types.append(ga.SIZE)  # k
+        param_types.append(ga.GpuArray)  # src
+        param_types.extend([ga.SSIZE] * ndim)  # src_strides
+        param_types.append(ga.SIZE)  # size
+
+        # load and compile kernels
+        with open(os.path.join(
+            os.path.dirname(__file__), 'c_code', 'k_topk_common' + common_ext
+        )) as f:
+            common_src = f.read()
+
+        kernels = []
+
+        def build_kernel(fname, kname, subs):
+            with open(os.path.join(
+                os.path.dirname(__file__), 'c_code', fname)
+            ) as f:
+                kernel_src = f.read()
+            ker = Kernel(
+                code=Template(common_src + kernel_src).substitute(**subs),
+                name=kname,
+                params=param_types,
+                flags=flags,
+                objvar=kname + nodename)
+            return ker
+
+        subs['count_t'] = 'int'
+        kernels.append(
+            build_kernel('k_topk_dense' + kernel_ext, 'k_topk_dense', subs))
+        subs['kname'] = 'k_topk_dense_large'
+        kernels.append(
+            build_kernel('k_topk_dense_large' + kernel_ext, 'k_topk_dense_large', subs))
+        subs['count_t'] = 'long long'
+        subs['kname'] = 'k_topk_dense_xlarge'
+        kernels.append(
+            build_kernel('k_topk_dense_large' + kernel_ext, 'k_topk_dense_xlarge', subs))
+        return kernels
+
+    def c_code(self, node, nodename, inps, outs, sub):
+        if node.inputs[0].type.context.kind != b'cuda':
+            raise NotImplementedError(
+                '%s: We only have CUDA '
+                'implementation so far.' % self.__class__.__name__)
+        x, k = inps
+        inp_dtc = ga.dtype_to_typecode(node.inputs[0].dtype)
+        if not self.return_indices:
+            yv, = outs
+        elif self.return_values:
+            yv, yi = outs
+        else:
+            yi, = outs
+        out_dtype_s = self.idx_dtype
+        out_dtc = ga.dtype_to_typecode(out_dtype_s)
+        fail = sub['fail']
+        ctx = sub['params']
+        k_dtype = node.inputs[1].type.dtype_specs()[1]
+        MAX_TPB = 1024  # max threads per block
+        WARP_SIZE = 32
```

Comment:
I would still hardcode `WARP_SIZE`, reasons:

1. The kernel code relies on the fact that warp size is constant 32. Making this value dynamic may drop performance and adding complexity.

2. This value is constant across all CUDA GPUs. AFAIK many code just assume it's constant 32. NVIDIA probably won't change it in short term.